### ISSUE TITLE
Support for different locales?

### ISSUE THREAD

**Issue by pdehaan** [mozilla/idea-town#48]
Speaking of error messages, etc.
Not sure if it makes sense to decide on some l10n scheme and extract all the strings from the beginning (rather than trying to add support for different languages in the future).

**Comment by johngruen** [comment]
@pdehaan I don't think this is a target for our ship date next week, but we should have a story here sooner rather than later.

**Comment by 6a68** [comment]
What's the fxa workflow for l10n? should we borrow it, or do something new?

**Comment by pdehaan** [comment]
I have no clue. I think every project I've worked on has implemented it differently.
FxA has a fxa-content-server-l10n repo where the localizers put translations (via verbatim, iirc) and then we grab that repo using bower and blah blah blah.

I defer to @nchapman here for how he wants to proceed (in Q4 or beyond).

**Comment by nchapman** [comment]
We used the fxa workflow for Find My Device, so I'm familiar with the steps necessary to get it going. It's a bit of a Rube Goldberg machine, but seems to work well enough once it's going.

**Comment by pdehaan** [comment]
jrgm and I were looking into http://l20n.org/ (https://github.com/l20n/l20n.js) last night and there was a discussion on #fxa earlier. It actually looks pretty interesting. Powered by Mozilla.

![](http://www.intermedia4web.com/wp-content/uploads/2012/11/eating-your-own-dog-food-lg.png)

**Comment by pdehaan** [comment]
Actually, here was my super crude idea for implementing it:

```html
<html>
  <head>
    <meta charset="utf-8">
    <title data-l10n-id="title"></title>

    <meta name="defaultLanguage" content="en-US">
    <meta name="availableLanguages" content="en-US, fr">
    <link rel="localization" href="locales/example.{locale}.l20n">
    <script
      type="application/javascript;version=1.8"
      defer src="../dist/web/l20n.js"></script>
  </head>

  <body>
    <h1 data-l10n-id="hello"></h1>
    <p data-l10n-id="name">
      <input type="text">
      <input type="submit">
    </p>
    <div data-l10n-id="unreadEmails" data-l10n-args='{"emailCount": 5}'/>
  </body>
</html>
```
&mdash; via https://github.com/l20n/l20n.js/blob/master/examples/web.html

So we apparently just define l20n string IDs as data attributes on certain elements. Interpolated values (like `emailCount`) seem a bit trickier, but probably doable in our Jade templates.
We could also pass in the `defaultLanguage` and `availableLanguages` in to our Jade rendering engine via **convict** configs (to make it easy to toggle different values locally or in stage/prod or via CLI).

Specifying the localization paths may be a bit trickier:
```html
<link rel="localization" href="locales/example.{locale}.l20n">
```

Either all those [translation files](https://github.com/l20n/l20n.js/blob/master/examples/locales/example.en-US.l20n) can live in our repo, or if we want a separate **idea-town-l10n** repo, it'd be easy enough to import those in via npm or bower at build time and use Gulp to shuffle the files around as needed.

**Comment by pdehaan** [comment]
WIP branch at https://github.com/mozilla/idea-town/tree/l20n
No doubt needs a dozen rebases already, but shows how the page will seamlessly update translations when you change your Firefox locale without a page refresh.

Still not sure how I can set a `data-l10n-id` attribute to set the placeholder text on an `<input type="email">` field, or how I could get this to place nice w/ @johngruen's arrays of strings in the [data.json](https://github.com/mozilla/idea-town/blob/master/routes/data.json) file.

So far I officially kind of like this l20n API. Put me down as a solid maybe!

**Comment by meandavejustice** [comment]
I20n looks relatively painless
